compiler/next/include/chpl/queries/stringify-functions.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This file implements the generic chpl::stringify
+  as well as specializations for some common types.
+
+  stringify can be used to get a string representation of an object
+  for debugging purposes
+ */
+
+#ifndef CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+#define CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "chpl/util/memory.h"
+#include <set>
+#include <sstream>
+
+
+namespace chpl {
+class Context;
+
+enum StringifyKind {
+  DEBUG_SUMMARY,
+  DEBUG_DETAIL,
+  CHPL_SYNTAX
+};
+
+
+template<typename T> struct stringify {
+  std::string operator()(StringifyKind stringKind, const T& stringMe) const = 0;
+};
+
+
+template<typename T> struct stringify<const T*> {
```

Comment:
suggest comment before: `// define stringify for pointers to call stringify on a reference to avoid duplication for types that appear both as references and as pointers`